src/actions/ast.js

```diff
@@ -53,17 +56,49 @@ export function setSourceMetaData(sourceId: SourceId) {
 
 export function setSymbols(sourceId: SourceId) {
   return async ({ dispatch, getState, sourceMaps }: ThunkArgs) => {
-    const source = getSourceFromId(getState(), sourceId);
+    const id = sourceId;
+    const source = getSourceFromId(getState(), id);
+    // Fetch the source text only once.
+    if (requests.has(id)) {
+      return requests.get(id);
+    }
+
+    if (isLoaded(source)) {
```

Comment:
hmm, i think we want to check if there are symbols, and if so return early

Reply:
@jasonLaster after checking id or instead of id we need to check for symbols and is there any function predefined

Reply:
i don't follow

Reply:
Is there any function defined somewhere which returns whether symbols have fetched or not?

Reply:
i think `hasSymbols`